python/lsst/afw/multiband.py

```diff
@@ -0,0 +1,463 @@
+# This file is part of afw.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ["MultibandBase"]
+
+import numpy as np
+
+from lsst.geom import Point2I, Box2I
+from .image import PARENT
+
+
+class MultibandBase(object):
+    """Base class for multiband objects
+
+    The LSST stack has a number of image-like classes that have
+    data in multiple bands that are stored as separate objects.
+    Analyzing the data can be easier using a Multiband object that
+    wraps the underlying data as a single data cube that can be sliced and
+    updated as a single object.
+
+    `MultibandBase` is designed to contain the most important universal
+    methods for initializing, slicing, and extracting common parameters
+    (such as the bounding box or XY0 position) to all of the single band classes.
+    """
+    def __init__(self, filters, singles):
+        """Initialize a `MultibandBase` object
+
+        Must be overloaded in derived classes to use `array` and slicing
+        functionality.
+
+        Parameters
+        ----------
+        filters: list
+            List of filter names.
+        singles: list
+            List of single band objects
+        """
+        self._filters = tuple([f for f in filters])
+        self._singles = tuple(singles)
+        self._bbox = self._singles[0].getBBox()
+        self._singleType = type(self._singles[0])
```

Comment:
I don't think so. In one case a `MultibandImage` must have `singleType = ImageF` while another instance might have `singleType=ImageI`, for example. The reason the the `singleType` is stored is because for images, it is really the 3d (filters, y, x) data array that is stored and when a multiband object is sliced, new instances of the single filter objects (like `ImageF`) are created that point to that array, so the data type matters. If `MultibandMask` didn't inherit from `MultibandImage` (for example) I could have just used dtype of the array and initialized with `Image(array, dtype=array.dtype)` but since some of the classes are inherited this seemed like the easiest way to always construct the proper single band objects.

If you have a better method I'm happy to use it.

Reply:
I actually did that originally, but after all of the methods sorted themselves out I had
```
class MultibandImage(MultibandImageBase):
    pass
```

which seemed silly. But if you think that would be useful then I don't mind removing `singleType` and just using the `dtype` to create new single band instances.

Reply:
No, I don't feel that strongly about it, and if you've already checked that my proposal doesn't help in other ways, it's probably best to leave as-is.